mountpoint-s3/src/prefetch.rs

```diff
@@ -458,61 +469,45 @@ where
         let total_seek_distance = offset - self.next_sequential_read_offset;
         histogram!("prefetch.seek_distance", "dir" => "forward").record(total_seek_distance as f64);
 
-        let Some(current_task) = self.current_task.as_mut() else {
+        let Some(task) = self.backpressure_task.as_mut() else {
             // Can't seek if there's no requests in flight at all
             return Ok(false);
         };
 
-        // Jump ahead to the right request
-        if offset >= current_task.end_offset() {
-            self.next_sequential_read_offset = current_task.end_offset();
-            self.current_task = None;
-            while let Some(next_request) = self.future_tasks.pop_front() {
-                if next_request.end_offset() > offset {
-                    self.current_task = Some(next_request);
-                    break;
-                } else {
-                    self.next_sequential_read_offset = next_request.end_offset();
-                }
-            }
-            if self.current_task.is_none() {
-                // No inflight task containing the target offset.
-                trace!(current_offset=?self.next_sequential_read_offset, requested_offset=?offset, "seek failed: not enough inflight data");
-                return Ok(false);
-            }
-            // We could try harder to preserve the backwards seek buffer if we're near the
-            // request boundary, but it's probably not worth the trouble.
-            self.backward_seek_window.clear();
-        }
-
-        // At this point it's guaranteed by the previous if-block that `offset` is in the range of `self.current_task`
-        let current_task = self
-            .current_task
-            .as_mut()
-            .expect("a request existed that covered this seek offset");
         // If we have enough bytes already downloaded (`available`) to skip straight to this read, then do
         // it. Otherwise, we're willing to wait for the bytes to download only if they're coming "soon", where
         // soon is defined as up to `max_forward_seek_wait_distance` bytes ahead of the available offset.
-        let available_offset = current_task.available_offset();
-        if offset >= available_offset.saturating_add(self.config.max_forward_seek_wait_distance) {
+        let available_offset = task.available_offset();
+        let available_soon_offset = available_offset.saturating_add(self.config.max_forward_seek_wait_distance);
+        if offset >= available_soon_offset {
             trace!(
                 requested_offset = offset,
                 available_offset = available_offset,
                 "seek failed: not enough data available"
             );
             return Ok(false);
         }
+
+        // Make sure there is enough data in the read window
```

Comment:
I think you're right, will update this.

Reply:
+1 for resetting if `next_read_window_offset < available_soon_offset`.

So `max_forward_seek_wait_distance` [today](https://github.com/awslabs/mountpoint-s3/blob/main/mountpoint-s3/src/prefetch.rs#L149) is `16MB` and we assume that this is more efficient for the existing request to download `16MB` instead of creating a new one. Monthon's code looks aligned with this assumption (after line [482](https://github.com/awslabs/mountpoint-s3/pull/926/files#diff-ff2b2b24fbf3d7ad046bb857a44d84da70bbdd714b0ea9460f5d1b8a08408effR482) we know that it's less than `16MB` to be downloaded left to serve the read).

I'd say the problem here is that we may increment the window exactly to the requested byte and no further. Any window incrementing rule which prevents this looks arbitrary or complicated, so I'd just reset too.